cpp/include/cudf/strings/detail/copy_range.cuh

```diff
@@ -0,0 +1,217 @@
+ */
+#pragma once
+
+#include <cudf/types.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+/**
+ * @brief Internal API to copy a range of string elements out-of-place from
+ * source iterators to a target column.
+ *
+ * Creates a new column as if an in-place copy was performed into @p target.
+ * The elements indicated by the indices [@p target_begin, @p target_end) were
+ * replaced with the elements retrieved from source iterators;
+ * *(@p source_value_begin + idx) if *(@p source_validity_begin + idx) is true,
+ * invalidate otherwise (where idx = [0, @p target_end - @p target_begin)).
+ * Elements outside the range are copied from @p target into the new target
+ * column to return.
+ *
+ * @throws `cudf::logic_error` for invalid range (if @p target_begin < 0,
+ * target_begin >= @p target.size(), or @p target_end > @p target.size()).
+ *
+ * @tparam SourceValueIterator Iterator for retrieving source values
+ * @tparam SourceValidityIterator Iterator for retrieving source validities
+ * @param source_value_begin Start of source value iterator
+ * @param source_validity_begin Start of source validity iterator
+ * @param target The strings column to copy from outside the range.
+ * @param target_begin The starting index of the target range (inclusive)
+ * @param target_end The index of the last element in the target range
+ * (exclusive)
+ * @param mr Memory resource to allocate the result target column.
+ * @param stream CUDA stream to run this function
+ * @return std::unique_ptr<column> The result target column
+ */
+template <typename SourceValueIterator, typename SourceValidityIterator>
+std::unique_ptr<column> copy_range(SourceValueIterator source_value_begin,
+                                   SourceValidityIterator source_validity_begin,
+                                   strings_column_view const& target,
+                                   // TODO: first & last are used for the fill
+                                   // function for strings (instead of begin &
+                                   // end)
+                                   size_type target_begin, size_type target_end,
+                                   rmm::mr::device_memory_resource* mr =
+                                     rmm::mr::get_default_resource(),
+                                   cudaStream_t stream = 0) {
+  CUDF_EXPECTS((target_begin >= 0) &&
+                 (target_begin < target.size()) &&
+                 (target_end <= target.size()),
+               "Range is out of bounds.");
+
+  if (target_end != target_begin) {
```

Comment:
I had to scroll way up to figure out that this is the `if` that corresponds to the `else` below:
```
  else {
    return std::make_unique<column>(target.parent(), stream, mr);
  }
```
I think this would be easier to follow. 

```suggestion
  if (target_end == target_begin) 
     return std::make_unique<column>(target.parent(), stream, mr);
```